File: Session_02_Recurrences/multi_language_implementations/recurrence_analysis.swift
import Foundation

func printIndent(_ depth: Int) {
    for _ in 0..<depth {
        print("  |  ", terminator: "")
    }
}

func calculateFactorial(_ n: Int, _ depth: Int) -> Int64 {
    printIndent(depth)
    var temp = n
    let address = withUnsafePointer(to: &temp) { "\($0)" }
    print("Entering factorial(\(n)). Address of n: \(address)")
    if n <= 1 {
        printIndent(depth)
        print("Base Case reached for factorial(\(n)). Returning 1.")
        return 1
    }
    let subResult = calculateFactorial(n - 1, depth + 1)
    let result = Int64(n) * subResult
    printIndent(depth)
    print("Exiting factorial(\(n)) -> returning \(n) * \(subResult) = \(result)")
    return result
}

func calculateFibonacci(_ n: Int, _ depth: Int, _ branch: String) -> Int {
    printIndent(depth)
    var temp = n
    let address = withUnsafePointer(to: &temp) { "\($0)" }
    print("Entering fibonacci(\(n)) [\(branch) branch]. Address of n: \(address)")
    if n == 0 {
        printIndent(depth)
        print("Base Case: fibonacci(0) = 0. Returning.")
        return 0
    }
    if n == 1 {
        printIndent(depth)
        print("Base Case: fibonacci(1) = 1. Returning.")
        return 1
    }
    let left = calculateFibonacci(n - 1, depth + 1, "Left")
    let right = calculateFibonacci(n - 2, depth + 1, "Right")
    let result = left + right
    printIndent(depth)
    print("Exiting fibonacci(\(n)) -> returning \(left) + \(right) = \(result)")
    return result
}

func main() {
    let val = 5
    print("=== Recursion & Stack Frame Visualizer ===\n")
    print("--- PART 1: Linear Recursion - Factorial(\(val)) ---")
    let fact = calculateFactorial(val, 0)
    print("\nFactorial of \(val) is: \(fact)\n")
    print("--------------------------------------------------\n")
    print("--- PART 2: Tree Recursion - Fibonacci(4) ---")
    let fib = calculateFibonacci(4, 0, "Root")
    print("\nFibonacci of 4 is: \(fib)\n")
    print("Press Enter to exit...", terminator: "")
    fflush(stdout)
    let _ = readLine()
}

main()

File: Session_02_Recurrences/multi_language_implementations/recurrence_analysis.py
import sys

def print_indent(depth):
    print("  |  " * depth, end="")

def get_address(depth):
    # Simulated stack address starting at 0x7ffebfaf1000 and growing down
    base = 0x7ffebfaf1000
    return f"0x{base - depth * 0x30:x}"

def calculate_factorial(n, depth):
    print_indent(depth)
    print(f"Entering factorial({n}). Address of n: {get_address(depth)}")
    if n <= 1:
        print_indent(depth)
        print(f"Base Case reached for factorial({n}). Returning 1.")
        return 1
    sub_result = calculate_factorial(n - 1, depth + 1)
    result = n * sub_result
    print_indent(depth)
    print(f"Exiting factorial({n}) -> returning {n} * {sub_result} = {result}")
    return result

def calculate_fibonacci(n, depth, branch):
    print_indent(depth)
    print(f"Entering fibonacci({n}) [{branch} branch]. Address of n: {get_address(depth)}")
    if n == 0:
        print_indent(depth)
        print("Base Case: fibonacci(0) = 0. Returning.")
        return 0
    if n == 1:
        print_indent(depth)
        print("Base Case: fibonacci(1) = 1. Returning.")
        return 1
    left = calculate_fibonacci(n - 1, depth + 1, "Left")
    right = calculate_fibonacci(n - 2, depth + 1, "Right")
    result = left + right
    print_indent(depth)
    print(f"Exiting fibonacci({n}) -> returning {left} + {right} = {result}")
    return result

def main():
    val = 5
    print("=== Recursion & Stack Frame Visualizer ===\n")
    print(f"--- PART 1: Linear Recursion - Factorial({val}) ---")
    fact = calculate_factorial(val, 0)
    print(f"\nFactorial of {val} is: {fact}\n")
    print("--------------------------------------------------\n")
    print("--- PART 2: Tree Recursion - Fibonacci(4) ---")
    fib = calculate_fibonacci(4, 0, "Root")
    print(f"\nFibonacci of 4 is: {fib}\n")
    input("Press Enter to exit...")

if __name__ == "__main__":
    main()
